### PR TITLE
update README for download/bulkupload flows

### PR DESCRIPTION
**Pull request by jeremymv2**
Closes  https://github.com/habitat-sh/habitat/issues/7015
Closes https://github.com/habitat-sh/habitat/issues/7018

This updates the top-level README with the new process for bootstrapping an on-prem Builder instance.

As a bonus, it introduces some very minor changes to the `install.sh` and `scripts/install-hab.sh` files to support Airgap installs, along with the preparation work necessary for an airgap install. The prep work leverages `hab pkg download` and the `quickstart_lists/builder_x86_64-linux_stable` file.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -140,6 +140,39 @@ sudo openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout /etc/ssl/privat
 
 *Important*: Make sure that the certificate files are named exactly `ssl-certificate.key` and `ssl-certificate.crt`. If you have procured the certificate from a different source, rename them to the prescribed filenames, and ensure that they are located in the same folder as the `install.sh` script. They will get uploaded to the Chef Habitat supervisor during the install.
 
+### Airgapped Installation (Recommended prep work, if airgapped)
```
Comment (chefsalim) [inline README.md]: I'm not sure we are ready to put in Airgap instructions yet until the installation actually supports it (ie, the install script runs and completes without an Internet connection?).

File: README.md [hunk 2]
```diff
@@ -151,12 +184,11 @@ sudo openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout /etc/ssl/privat
 If everything goes well, you should see output similar to the following showing that the Chef Habitat Builder on-prem services are loaded:
 
 ```output
-hab-sup(MN): The habitat/builder-datastore service was successfully loaded
-hab-sup(MN): The habitat/builder-router service was successfully loaded
-hab-sup(MN): The habitat/builder-api service was successfully loaded
-hab-sup(MN): The habitat/builder-api-proxy service was successfully loaded
-hab-sup(MN): The habitat/builder-originsrv service was successfully loaded
-hab-sup(MN): The habitat/builder-sessionsrv service was successfully loaded
+hab-sup(AG): The habitat/builder-datastore service was successfully loaded
```
Comment (chefsalim) [inline README.md]: Why these changes?

File: README.md [hunk 3]
```diff
@@ -187,46 +219,32 @@ Next, generate a Personal Access Token for bootstrapping the `core` packages, as
...
 
+1. Phase 1: download
     ```bash
-    sudo -E ./scripts/on-prem-archive.sh populate-depot http://${BUILDER_HOSTNAME_OR_IP}`
+    export HAB_AUTH_TOKEN=<your _public_ Builder instance token>
+    cd on-prem-builder
+    hab pkg download --target x86_64-linux --channel stable --file quickstart_lists/effortless_x86_64-linux_stable --download-directory builder_bootstrap
+    ```
+1. Phase 2: bulkupload
+*Important*: Inspect the contents of the `builder_bootstrapi/artifacts` directory created from the download command above. For each of the origins (`core`, `effortless`, etc),  create a matching origin name if one doesn't exist already in the on-prem Builder UI before starting the bulkupload.
```
Comment (chefsalim) [inline README.md]: Typo - "builder_bootstrapi"

Comment (chefsalim) [reply README.md]: Change "create a matching origin name" to "create the origin"

File: README.md [hunk 4]
```diff
@@ -187,46 +219,32 @@ Next, generate a Personal Access Token for bootstrapping the `core` packages, as
...
+    ```bash
+    export HAB_AUTH_TOKEN=<your _on-prem_ Builder instance token>
```
Comment (chefsalim) [inline README.md]: Remove extra space

File: README.md [hunk 5]
```diff
@@ -187,46 +219,32 @@ Next, generate a Personal Access Token for bootstrapping the `core` packages, as
...
+    hab pkg bulkupload --url https://your-builder.tld --channel stable builder_bootstrap/
     ```
```
Comment (chefsalim) [inline README.md]: Instead of completely removing the Synchronization section, we should still have the section topic, but change the prescriptive guidance to use the download/bulkupload commands

File: README.md (hunk 4 above)
Comment (jeremymv2) [reply README.md]: Handled in https://github.com/habitat-sh/on-prem-builder/pull/212/commits/634b062e33a6a3d4aede2d6788fc9bf0f4160152

File: README.md (hunk 5 above)
Comment (jeremymv2) [reply README.md]: Handled in https://github.com/habitat-sh/on-prem-builder/pull/212/commits/634b062e33a6a3d4aede2d6788fc9bf0f4160152

I re-added the section with a deprecation notice:

## Bootstrap `core` packages (**Deprecated**)

File: README.md (hunk 3 above)
Comment (jeremymv2) [reply README.md]: Handled in https://github.com/habitat-sh/on-prem-builder/pull/212/commits/634b062e33a6a3d4aede2d6788fc9bf0f4160152

File: README.md (hunk 2 above)
Comment (jeremymv2) [reply README.md]: Oh, this change is simply due to the interesting fact that I am not seeing those same service names in my local Builder.

Specifically, I don't see `builder-originsrv` or `habitat/builder-sessionsrv`.
Thus, I based the changes off of the list of services that I do see with `hab svc status` and `journalctl | grep loaded`:

```
root@chef-builder:~# hab svc status
package                                        type        desired  state  elapsed (s)  pid   group
habitat/builder-api/8483/20190917211441        standalone  up       up     255440       7871  builder-api.default
habitat/builder-datastore/7795/20181018210336  standalone  up       up     255441       7800  builder-datastore.default
habitat/builder-minio/7764/20181006010221      standalone  up       up     255441       7820  builder-minio.default
habitat/builder-api-proxy/8496/20190924062940  standalone  up       up     255441       7841  builder-api-proxy.default
habitat/builder-memcached/7728/20180929144821  standalone  up       up     255441       7851  builder-memcached.default
root@chef-builder:~#
```

Perhaps the previous list was obsoleted?

File: README.md (hunk 1 above)
Comment (kagarmoe) [reply README.md]: Our file structure is pretty complicated. I think it is time for me to add a local documentation server & structure. Let's keep the airgap here for _today_. Once merged, I can put it onto a draft page.

Comment (jeremymv2) [reply README.md]: @chefsalim I'm very pleased to assure you that the changes in this PR accomplish the following:
* Preparation work enumerated in detail for an Airgap Installation.
* Once prep-work is completed, the Airgap install with no Internet connection, has been validated multiple times to be working successfully.
* Install _with_ Internet connection has been validated to work just as before.

File: builder-automate.md [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+
+```bash
+bash ./install.sh
+```
+
+Restart the Chef Habitat services:
+
+```bash
+sudo systemctl restart hab-sup
+```
```
Comment (jeremymv2) [inline builder-automate.md]: Hi @kagarmoe! I'm not completely certain why we're adding this document in my PR?  While I seem to agree with the file's contents, it feels a little out of scope from the intent of my PR.

I just don't want it to get in the way of the team's approval. If the team approves of this, I'm fine to leave this in.
Thank you!

Comment (kagarmoe) [reply builder-automate.md]: Sorry, an unaffiliated file got pushed when I corrected a few minor markdown formatting issues.

Comment (jeremymv2) [reply builder-automate.md]: No worries. Thanks @kagarmoe!

File: README.md (hunk 1 above)
Comment (chefsalim) [reply README.md]: Excellent, that is great to know @jeremymv2!

File: README.md (hunk 2 above)
Comment (chefsalim) [reply README.md]: Ah yes, that makes sense now, thanks

File: README.md [hunk 7]
```diff
@@ -140,23 +140,77 @@ sudo openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout /etc/ssl/privat
...
+    curl -LO https://github.com/habitat-sh/on-prem-builder/archive/master.zip
+    ```
+
+1. Download the Chef Habitat [cli tool](https://api.bintray.com/content/habitat/stable/linux/x86_64/hab-%24latest-x86_64-linux.tar.gz?bt_package=hab-x86_64-linux)
+
+    ```bash
+    curl -Lo hab.tar.gz https://api.bintray.com/content/habitat/stable/linux/x86_64/hab-%24latest-x86_64-linux.tar.gz?bt_package=hab-x86_64-linux
+    ```
+
+1. Create the Habitat Builder starter kit bundle and download it
```
Comment (chefsalim) [inline README.md]: "starter kit" - should reflect the final terminology we land on

File: README.md [hunk 8]
```diff
@@ -140,23 +140,77 @@ sudo openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout /etc/ssl/privat
...
+
+     ```bash
+     git clone https://github.com/habitat-sh/on-prem-builder.git
+     export DOWNLOAD_DIR=/some/base/download/directory
+     cd on-prem-builder
+     hab pkg download --target x86_64-linux --channel stable --file quickstart_lists/builder_x86_64-linux_stable --download-directory ${DOWNLOAD_DIR}/builder_starter_kit
+     ```
+
+1. Create any additional starter kit Builder bootstrap bundles as documented in the [Bootstrap Builder](https://github.com/habitat-sh/on-prem-builder/tree/master#bootstrap-builder-with-habitat-packages) section of this README. You can specify `--download-directory ${DOWNLOAD_DIR}/builder_bootstrap` argument to the download command in order to consolidate all bootstrap packages in a single directory
```
Comment (chefsalim) [inline README.md]: "starter kit" - replace with final terminology we land on

File: README.md [hunk 9]
```diff
@@ -140,23 +140,77 @@ sudo openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout /etc/ssl/privat
 
 *Important*: Make sure that the certificate files are named exactly `ssl-certificate.key` and `ssl-certificate.crt`. If you have procured the certificate from a different source, rename them to the prescribed filenames, and ensure that they are located in the same folder as the `install.sh` script. They will get uploaded to the Chef Habitat supervisor during the install.
 
+### Airgapped Installation (Recommended prep work, if airgapped)
+
+In order to install the on-prem Chef Habitat Builder in an airgapped (no direct Internet access) environment, the following preparatory steps are helpful:
```
Comment (chefsalim) [inline README.md]: May be obvious, but worth calling out that the prep steps are to be done in a separate non-airgapped environment?

File: README.md [hunk 10]
```diff
@@ -140,23 +140,77 @@ sudo openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout /etc/ssl/privat
...
+1. Create any additional starter kit Builder bootstrap bundles as documented in the [Bootstrap Builder](https://github.com/habitat-sh/on-prem-builder/tree/master#bootstrap-builder-with-habitat-packages) section of this README. You can specify `--download-directory ${DOWNLOAD_DIR}/builder_bootstrap` argument to the download command in order to consolidate all bootstrap packages in a single directory
+1. Zip up all the above, transfer and unzip on the Linux system where Builder will be deployed
+1. From the zip archive, install the `hab` binary somewhere in $PATH and ensure it has execute permissions:
+
+     ```bash
+     sudo chmod 755 /usr/bin/hab
+     sudo hab # read the license and accept if in agreement, as the root user
+     ```
+
+1. Import the public package signing keys from the downloaded Builder starter kit:
```
Comment (chefsalim) [inline README.md]: "starter kit" - replace with final terminology

File: README.md [hunk 11]
```diff
@@ -140,23 +140,77 @@ sudo openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout /etc/ssl/privat
...
+     ```bash
+     git clone https://github.com/habitat-sh/on-prem-builder.git
+     export DOWNLOAD_DIR=/some/base/download/directory
+     cd on-prem-builder
+     hab pkg download --target x86_64-linux --channel stable --file quickstart_lists/builder_x86_64-linux_stable --download-directory ${DOWNLOAD_DIR}/builder_starter_kit
+     ```
+
+1. Create any additional starter kit Builder bootstrap bundles as documented in the [Bootstrap Builder](https://github.com/habitat-sh/on-prem-builder/tree/master#bootstrap-builder-with-habitat-packages) section of this README. You can specify `--download-directory ${DOWNLOAD_DIR}/builder_bootstrap` argument to the download command in order to consolidate all bootstrap packages in a single directory
+1. Zip up all the above, transfer and unzip on the Linux system where Builder will be deployed
+1. From the zip archive, install the `hab` binary somewhere in $PATH and ensure it has execute permissions:
```
Comment (chefsalim) [inline README.md]: Maybe separate this part into its own subsection to differentiate the work that has to be done in airgap prep, vs on the airgap environment

File: README.md [hunk 12]
```diff
@@ -140,23 +140,77 @@ sudo openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout /etc/ssl/privat
...
+     sudo hab pkg install /hab/cache/artifacts/habitat-builder*hart
+     ```
+
+1. Pre-install the Habitat Supervisor and its dependencies:
+
+     ```bash
+     sudo hab pkg install --binlink --force /hab/cache/artifacts/core-hab-*hart
+     ```
+
 ## Setup
```
Comment (chefsalim) [inline README.md]: Should this now have "Airgapped" and "Non-Airgapped" setup sub sections?

File: README.md [hunk 13]
```diff
@@ -292,32 +379,29 @@ Restoring a `builder` database is exactly like restoring any other database--whi
         `/hab/pkgs/core/postgresql/<version>/<release>/bin/pg_restore --host=<url_of_pg_host> --dbname=builder builder.dump`
 1. Start the on-prem Builder services
 
-        * In some cases your version of Postgres might not have a createdb binary in which case you'll want to connect to database to run the create db command
+    > note: In some cases your version of Postgres might not have a createdb binary in which case you'll want to connect to database to run the create db command.
```
Comment (chefsalim) [inline README.md]: Capitalize "note"?

Comment (jeremymv2) [reply README.md]: Addressed in https://github.com/habitat-sh/on-prem-builder/pull/212/commits/38454a460a0f673bf5f0cde08392dfcb3e36b1f6

File: README.md (hunk 12 above)
Comment (jeremymv2) [reply README.md]: Addressed in https://github.com/habitat-sh/on-prem-builder/pull/212/commits/38454a460a0f673bf5f0cde08392dfcb3e36b1f6

File: README.md (hunk 11 above)
Comment (jeremymv2) [reply README.md]: Addressed in https://github.com/habitat-sh/on-prem-builder/pull/212/commits/38454a460a0f673bf5f0cde08392dfcb3e36b1f6

File: README.md (hunk 9 above)
Comment (jeremymv2) [reply README.md]: Addressed in https://github.com/habitat-sh/on-prem-builder/pull/212/commits/38454a460a0f673bf5f0cde08392dfcb3e36b1f6

File: README.md (hunk 10 above)
Comment (jeremymv2) [reply README.md]: Handled in https://github.com/habitat-sh/on-prem-builder/pull/212/commits/5e3e473c99054abadcd0dbd9bfd3cb66e9f41548 "package seed list"

File: README.md (hunk 8 above)
Comment (jeremymv2) [reply README.md]: Handled in https://github.com/habitat-sh/on-prem-builder/pull/212/commits/5e3e473c99054abadcd0dbd9bfd3cb66e9f41548 "package seed list"

File: README.md (hunk 7 above)
Comment (jeremymv2) [reply README.md]: Handled in https://github.com/habitat-sh/on-prem-builder/pull/212/commits/5e3e473c99054abadcd0dbd9bfd3cb66e9f41548 "package seed list"

File: README.md [hunk 14]
```diff
@@ -140,23 +140,86 @@ sudo openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout /etc/ssl/privat
...
+     git clone https://github.com/habitat-sh/on-prem-builder.git
+     export DOWNLOAD_DIR=/some/base/download/directory
+     cd on-prem-builder
+     hab pkg download --target x86_64-linux --channel stable --file package_seed_lists/builder_x86_64-linux_stable --download-directory ${DOWNLOAD_DIR}/builder_packages
+     ```
+
+1. Create any additional package bundles to upload to Builder from package seed lists as documented in the [Bootstrap Builder](https://github.com/habitat-sh/on-prem-builder/blob/master/README.md#bootstrap-builder-with-habitat-packages-new) section of this README. You can specify `--download-directory ${DOWNLOAD_DIR}/builder_bootstrap` argument to the download command in order to consolidate all bootstrap packages in a single directory
+1. Zip up all the above content, transfer and unzip on the Linux system where Builder will be deployed in the Airgapped environment
+
+> Note: The following tasks are intended to be completed on the Airgapped system where Builder will be deployed, in advance of the installation
```
Comment (chefsalim) [inline README.md]: Should this section be part of the "Installation" section below, as this is actually the airgap "install", correct?

File: README.md [hunk 15]
```diff
@@ -140,23 +140,86 @@ sudo openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout /etc/ssl/privat
...
+
 ## Setup
 
-1. Clone this repo (or unzip the archive you have downloaded from the Github release page) at the desired machine where you will stand up the Chef Habitat Builder on-prem
+1. Clone this repo (or unzip the zip archive you previously downloaded from the Github release page) at the desired machine where you will stand up the Chef Habitat Builder on-prem
 1. `cd ${SRC_ROOT}`
 1. `cp bldr.env.sample bldr.env`
 1. Edit `bldr.env` with a text editor and replace the values appropriately. Consider helping us to improve Chef Habitat as well by changing the `ANALYTICS_ENABLED` setting to `true` and providing an optional company name.
+
+## Installation
```
Comment (chefsalim) [inline README.md]: Does this section apply for both airgap as well as non-airgap?

File: README.md [hunk 16]
```diff
@@ -187,7 +250,41 @@ Next, generate a Personal Access Token for bootstrapping the `core` packages, as
...
-## Bootstrap `core` packages
+## Bootstrap Builder with Habitat Packages (**New**)
+
+Chef Habitat Builder on-prem has no pre-installed package sets. You must populate your Builder instance by uploading packages.
+With Habitat *0.88.0*, two new commands were introduced to assist in bootstrapping an on-prem Builder instance with a set of stable packages:
+
+1. *hab pkg download*
+1. *hab pkg bulkupload*
+
+As you can see from the commands above, the package Bootstrap flow is comprised of two main phases: a download from the public [SaaS Builder](https://bldr.habitat.sh) followed by a bulkupload to your on-prem Builder instance(s). Historically, we bootstraped on-prem-builders by downloading all the packages in 'core' for all targets. That amounted to ~15GB and was both too much and too little, in that many of the packages weren't needed, and for many patterns (effortless) other origins were needed.
```
Comment (chefsalim) [inline README.md]: Typo: "bootstraped"

File: README.md (hunk 15 above)
Comment (jeremymv2) [reply README.md]: Yes, it is the same for airgap and non-airgap.

File: README.md [hunk 17]
```diff
@@ -187,7 +250,41 @@ Next, generate a Personal Access Token for bootstrapping the `core` packages, as
...
+     **Important**: Inspect the contents of the `builder_bootstrap/artifacts` directory created from the download command above. For each of the origins (`core`, `effortless`, etc),  create the origin name if one doesn't exist already in the on-prem Builder UI before starting the bulkupload.
+
+    ```bash
+    export HAB_AUTH_TOKEN=<your_on-prem_Builder_instance_token>
+    hab pkg bulkupload --url https://your-builder.tld --channel stable builder_bootstrap/
+    ```
+
+## Bootstrap `core` packages (**Deprecated**)
+
+*Important*: This methodology is now deprecated in favor of the download/bulkupload flow described above.
```
Comment (chefsalim) [inline README.md]: Also add the deprecation notice line to the "Synchronizing" section below?

File: README.md (hunk 16 above)
Comment (jeremymv2) [reply README.md]: Fixed in https://github.com/habitat-sh/on-prem-builder/pull/212/commits/bf14ce0f10ff40c35f15c48f6122f570929a452d

File: README.md (hunk 17 above)
Comment (jeremymv2) [reply README.md]: Agreed. Added in https://github.com/habitat-sh/on-prem-builder/pull/212/commits/bf14ce0f10ff40c35f15c48f6122f570929a452d

File: README.md (hunk 14 above)
Comment (jeremymv2) [reply README.md]: To me it feels more "preparation" work, whereas the install itself is just intuitively running the `./install.sh` script. If this prep work was done in advance, the `install.sh` is the same unifying experience irregardless.

To help strengthen that stance, I've renamed the "Preparatory Tasks for Airgapped ..." to "Prerequisite Tasks for an Airgapped Installation (Required if applicable)" and put a note in the Installation section that reads:

> Note: If the on-prem Builder system is in an Airgapped (non-Internet connected) environment, you must first complete the prerequisite tasks detailed earlier.

Take a look and see what you think of the flow from a new reader's perspective:

https://github.com/habitat-sh/on-prem-builder/tree/jeremymv2/airgap_and_download_upload#prerequisite-tasks-for-an-airgapped-installation-required-if-applicable 

Is that an improvement that addresses your concern adequately?

**Comment by jeremymv2** [comment]
@chefsalim and others.. I have also added a Note on SSL certchain requirements when the on-prem Builder has a non Publicly trusted CA issued SSL cert.  https://github.com/habitat-sh/on-prem-builder/pull/212/commits/8ee8f98ff6d37d8e4c304e0c3b38697173f9b54a

File: README.md (hunk 14 above)
Comment (chefsalim) [reply README.md]: Cool, that looks good.
